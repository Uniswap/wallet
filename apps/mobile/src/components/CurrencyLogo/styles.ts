import { ShadowProps } from '@shopify/restyle'
import { ImageResizeMode, StyleSheet } from 'react-native'
import { Theme } from 'src/styles/theme'

const RESIZE_MODE_CONTAIN: ImageResizeMode = 'contain'

export const style = StyleSheet.create({
  image: {
    resizeMode: RESIZE_MODE_CONTAIN,
  },
  innerWrapper: {
    borderRadius: 6,
  },
})

export const SHADOW_OFFSET: ShadowProps<Theme>['shadowOffset'] = { width: 0, height: 2 }

export const THIN_BORDER = 0.5
