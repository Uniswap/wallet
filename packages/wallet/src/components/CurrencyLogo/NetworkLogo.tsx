import React from 'react'
import { Image, StyleSheet } from 'react-native'
import { Box, BoxProps } from 'ui/src/components/layout/Box'
import { iconSizes } from 'ui/src/theme/iconSizes'
import { ChainId, CHAIN_INFO } from 'wallet/src/constants/chains'

import { ShadowProps } from '@shopify/restyle'
import { ImageResizeMode } from 'react-native'
import { Theme } from 'ui/src/theme/restyle/theme'

type NetworkLogoProps = BoxProps & {
  chainId: ChainId
  size?: number
  shape?: 'circle' | 'square'
}

export const SQUARE_BORDER_RADIUS = 6

export function TransactionSummaryNetworkLogo({
  chainId,
  size = iconSizes.icon20,
}: Pick<NetworkLogoProps, 'chainId' | 'size'>): JSX.Element {
  return (
    <Box borderColor="$background0" style={styles.squareLogoOutline}>
      <NetworkLogo chainId={chainId} shape="square" size={size} />
    </Box>
  )
}

function _NetworkLogo({
  chainId,
  shape = 'circle',
  size = iconSizes.icon20,
}: NetworkLogoProps): JSX.Element | null {
  const { logo } = CHAIN_INFO[chainId]
  const borderRadius = shape === 'circle' ? size / 2 : SQUARE_BORDER_RADIUS
  return logo ? (
    <Box style={[{ borderRadius }, styles.iconWrapper]}>
      <Image source={logo} style={[style.image, { width: size, height: size }]} />
    </Box>
  ) : null
}

export const NetworkLogo = React.memo(_NetworkLogo)

const styles = StyleSheet.create({
  iconWrapper: {
    overflow: 'hidden',
  },
  squareLogoOutline: {
    borderRadius: 7, // when inner icon borderRadius = 6 and size = 20
    borderWidth: 2,
  },
})

const RESIZE_MODE_CONTAIN: ImageResizeMode = 'contain'

export const style = StyleSheet.create({
  image: {
    resizeMode: RESIZE_MODE_CONTAIN,
  },
})

export const SHADOW_OFFSET: ShadowProps<Theme>['shadowOffset'] = { width: 0, height: 2 }

export const THIN_BORDER = 0.5
