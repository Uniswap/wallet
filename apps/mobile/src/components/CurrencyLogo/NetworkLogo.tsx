import { SpacingProps, SpacingShorthandProps } from '@shopify/restyle'
import React from 'react'
import { Image } from 'react-native'
import { style } from 'src/components/CurrencyLogo/styles'
import { Box } from 'src/components/layout/Box'
import { ChainId, CHAIN_INFO } from 'src/constants/chains'
import { iconSizes } from 'src/styles/sizing'
import { Theme } from 'src/styles/theme'
import { getNetworkColorKey } from 'src/utils/colors'

type NetworkLogoProps = {
  chainId: ChainId
  size?: number
} & SpacingProps<Theme> &
  SpacingShorthandProps<Theme>

export function TransactionSummaryNetworkLogo({
  chainId,
  size = iconSizes.icon20,
}: Pick<NetworkLogoProps, 'chainId' | 'size'>): JSX.Element {
  const { logo } = CHAIN_INFO[chainId]
  const backgroundColor = getNetworkColorKey(chainId)

  // We need to wrap the network logo because it's a circle, but we need a rectangle,
  // but we can use border and background color on the same view, as for some reason
  // background color is visible through the border on the outline
  // Similiar issue: https://stackoverflow.com/questions/43756623/react-native-border-radius-makes-outline
  return (
    <Box borderColor="background0" borderRadius="rounded8" borderWidth={2}>
      <Box backgroundColor={backgroundColor} borderColor="background0" style={style.innerWrapper}>
        {logo && (
          <Image
            source={logo}
            style={[
              style.image,
              {
                width: size,
                height: size,
              },
            ]}
          />
        )}
      </Box>
    </Box>
  )
}

export function NetworkLogo({
  chainId,
  size = iconSizes.icon20,
  ...rest
}: NetworkLogoProps): JSX.Element {
  const { logo } = CHAIN_INFO[chainId]

  return (
    <Box borderRadius="roundedFull" {...rest}>
      {logo && (
        <Image
          source={logo}
          style={[
            style.image,
            {
              width: size,
              height: size,
              borderRadius: size / 2,
            },
          ]}
        />
      )}
    </Box>
  )
}
