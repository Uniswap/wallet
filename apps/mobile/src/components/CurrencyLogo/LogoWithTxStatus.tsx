import React, { useEffect } from 'react'
import { StyleSheet } from 'react-native'
import { SvgProps } from 'react-native-svg'
import { useAppTheme } from 'src/app/hooks'
import AlertTriangle from 'src/assets/icons/alert-triangle.svg'
import Approve from 'src/assets/icons/approve.svg'
import IncomingArrow from 'src/assets/icons/arrow-down-in-circle.svg'
import OutgoingArrow from 'src/assets/icons/arrow-up-in-circle.svg'
import UnknownStatus from 'src/assets/icons/question-in-circle.svg'
import WalletConnectLogo from 'src/assets/icons/walletconnect.svg'
import MoonpayLogo from 'src/assets/logos/moonpay.svg'
import { CurrencyLogo, STATUS_RATIO } from 'src/components/CurrencyLogo'
import { TransactionSummaryNetworkLogo } from 'src/components/CurrencyLogo/NetworkLogo'
import { ImageUri } from 'src/components/images/ImageUri'
import { NFTViewer } from 'src/components/images/NFTViewer'
import { RemoteImage } from 'src/components/images/RemoteImage'
import { Box } from 'src/components/layout/Box'
import { DappIconPlaceholder } from 'src/components/WalletConnect/DappHeaderIcon'
import { ChainId } from 'src/constants/chains'
import { AssetType } from 'src/entities/assets'
import { CurrencyInfo } from 'src/features/dataApi/types'
import { NFTTradeType, TransactionStatus, TransactionType } from 'src/features/transactions/types'
import { WalletConnectEvent } from 'src/features/walletConnect/saga'
import { logger } from 'src/utils/logger'

interface LogoWithTxStatusProps {
  assetType: AssetType
  txType: TransactionType
  txStatus: TransactionStatus
  size: number
  chainId: ChainId | null
}

interface DappLogoWithTxStatusProps {
  event: WalletConnectEvent
  size: number
  chainId: ChainId | null
  dappImageUrl: string | null
  dappName: string
}

interface CurrencyStatusProps extends LogoWithTxStatusProps {
  assetType: AssetType.Currency
  currencyInfo?: CurrencyInfo | null
}

interface NFTStatusProps extends LogoWithTxStatusProps {
  assetType: AssetType.ERC721 | AssetType.ERC1155
  nftImageUrl?: string
  nftTradeType?: NFTTradeType
}

export function LogoWithTxStatus(props: CurrencyStatusProps | NFTStatusProps): JSX.Element {
  const { assetType, txType, txStatus, size, chainId } = props
  const theme = useAppTheme()

  const statusSize = size / 2

  const logo =
    txType === TransactionType.FiatPurchase ? (
      <MoonpayLogo width={size} />
    ) : assetType === AssetType.Currency ? (
      <CurrencyLogo hideNetworkLogo currencyInfo={props.currencyInfo} size={size} />
    ) : (
      <Box
        alignItems="center"
        backgroundColor="background2"
        borderRadius="rounded4"
        height={size}
        justifyContent="center"
        overflow="hidden"
        width={size}>
        {props.nftImageUrl && <NFTViewer uri={props.nftImageUrl} />}
      </Box>
    )

  const fill =
    txStatus === TransactionStatus.Success ? theme.colors.accentSuccess : theme.colors.textSecondary
  const color = theme.colors.background1

  let icon: JSX.Element | undefined
  if (chainId && chainId !== ChainId.Mainnet) {
    icon = <TransactionSummaryNetworkLogo chainId={chainId} size={size * STATUS_RATIO} />
  } else {
    let Icon: React.FC<SvgProps> | undefined
    switch (txType) {
      case TransactionType.Approve:
      case TransactionType.NFTApprove:
        Icon = Approve
        break
      case TransactionType.Send:
        Icon = OutgoingArrow
        break
      case TransactionType.NFTTrade:
        if (assetType === AssetType.ERC721 || assetType === AssetType.ERC1155) {
          if (props.nftTradeType === NFTTradeType.SELL) {
            Icon = OutgoingArrow
          } else {
            Icon = IncomingArrow
          }
        }
        break
      // Fiat purchases use the same icon as receive
      case TransactionType.FiatPurchase:
      case TransactionType.Receive:
      case TransactionType.NFTMint:
        Icon = IncomingArrow
        break
      case TransactionType.Unknown:
        Icon = UnknownStatus
        break
    }
    if (Icon) {
      icon = <Icon color={color} fill={fill} height={statusSize} width={statusSize} />
    }
  }

  useEffect(() => {
    if (!icon) {
      logger.warn(
        'statusIcon',
        'GenerateStatusIcon',
        'Could not find icon for transaction type:',
        txType
      )
    }
  }, [icon, txType])

  return (
    <Box alignItems="center" height={size} justifyContent="center" width={size}>
      {logo}
      {icon && (
        <Box bottom={-4} position="absolute" right={-4}>
          {icon}
        </Box>
      )}
    </Box>
  )
}

export function DappLogoWithTxStatus({
  dappName,
  dappImageUrl,
  event,
  size,
  chainId,
}: DappLogoWithTxStatusProps): JSX.Element | null {
  const theme = useAppTheme()
  const green = theme.colors.accentSuccess
  const yellow = theme.colors.accentWarning
  const fill = theme.colors.background0

  const dappImageSize = size
  const statusSize = dappImageSize * (1 / 2)
  const totalSize = dappImageSize + statusSize * (1 / 4)

  const getStatusIcon = (): JSX.Element | undefined => {
    switch (event) {
      case WalletConnectEvent.NetworkChanged:
        return chainId ? (
          <TransactionSummaryNetworkLogo chainId={chainId} size={statusSize} />
        ) : undefined
      case WalletConnectEvent.TransactionConfirmed:
        return <Approve color={green} fill={fill} height={statusSize} width={statusSize} />
      case WalletConnectEvent.TransactionFailed:
        return <AlertTriangle color={yellow} fill={fill} height={statusSize} width={statusSize} />
    }
  }

  const statusIcon = getStatusIcon()

  const fallback = (
    <Box height={dappImageSize}>
      <DappIconPlaceholder iconSize={dappImageSize} name={dappName} />
    </Box>
  )

  const style = StyleSheet.create({
    icon: {
      borderRadius: theme.borderRadii.rounded4,
      height: dappImageSize,
      width: dappImageSize,
    },
    loaderContainer: {
      borderRadius: theme.borderRadii.roundedFull,
      overflow: 'hidden',
    },
  })

  const dappImage = dappImageUrl ? (
    <ImageUri
      fallback={fallback}
      imageStyle={style.icon}
      loadingContainerStyle={{ ...style.icon, ...style.loaderContainer }}
      uri={dappImageUrl}
    />
  ) : (
    fallback
  )

  return statusIcon ? (
    <Box height={totalSize} width={totalSize}>
      <Box left={0} position="absolute" top={0}>
        {dappImage}
      </Box>
      <Box bottom={0} position="absolute" right={0}>
        {statusIcon}
      </Box>
    </Box>
  ) : (
    dappImage
  )
}

/** For displaying Dapp logo with generic WC bade icon */
export function DappLogoWithWCBadge({
  dappImageUrl,
  dappName,
  size,
  chainId,
}: {
  dappImageUrl: string | null
  dappName: string
  size: number
  chainId: ChainId | null
}): JSX.Element {
  const theme = useAppTheme()
  const dappImageSize = size
  const statusSize = dappImageSize * STATUS_RATIO
  const totalSize = dappImageSize + statusSize * (1 / 4)
  const dappImage = dappImageUrl ? (
    <RemoteImage
      borderRadius={theme.borderRadii.rounded4}
      height={dappImageSize}
      uri={dappImageUrl}
      width={dappImageSize}
    />
  ) : (
    <DappIconPlaceholder iconSize={dappImageSize} name={dappName} />
  )

  return (
    <Box height={totalSize} width={totalSize}>
      <Box left={2} top={0}>
        {dappImage}
      </Box>
      {chainId && chainId !== ChainId.Mainnet ? (
        <Box bottom={-2} position="absolute" right={-2}>
          <TransactionSummaryNetworkLogo chainId={chainId} size={size * STATUS_RATIO} />
        </Box>
      ) : (
        <Box
          backgroundColor="background1"
          borderColor="background0"
          borderRadius="roundedFull"
          borderWidth={2}
          bottom={0}
          position="absolute"
          right={0}>
          <WalletConnectLogo height={statusSize} width={statusSize} />
        </Box>
      )}
    </Box>
  )
}
