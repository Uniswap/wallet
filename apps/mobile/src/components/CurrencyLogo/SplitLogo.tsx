import React from 'react'
import { CurrencyLogo, STATUS_RATIO } from 'src/components/CurrencyLogo'
import { TransactionSummaryNetworkLogo } from 'src/components/CurrencyLogo/NetworkLogo'
import { Box } from 'src/components/layout/Box'
import { ChainId } from 'src/constants/chains'
import { CurrencyInfo } from 'src/features/dataApi/types'

interface Props {
  inputCurrencyInfo: NullUndefined<CurrencyInfo>
  outputCurrencyInfo: NullUndefined<CurrencyInfo>
  size: number
  chainId: ChainId | null
}

/*
 * Logo, where left 50% of width is taken from one icon (its left 50%)
 * and right side is taken from another icon (its right 50%)
 */
export function SplitLogo({
  size,
  inputCurrencyInfo,
  outputCurrencyInfo,
  chainId,
}: Props): JSX.Element {
  const iconSize = size / 2

  const icon =
    chainId && chainId !== ChainId.Mainnet ? (
      <TransactionSummaryNetworkLogo chainId={chainId} size={size * STATUS_RATIO} />
    ) : undefined

  return (
    <Box height={size} width={size}>
      <Box
        left={0}
        overflow="hidden"
        position="absolute"
        top={0}
        width={iconSize - 1 /* -1 to allow for space between the icons */}>
        <CurrencyLogo hideNetworkLogo currencyInfo={inputCurrencyInfo} size={size} />
      </Box>
      <Box
        flexDirection="row-reverse"
        overflow="hidden"
        position="absolute"
        right={0}
        top={0}
        width={iconSize - 1 /* -1 to allow for space between the icons */}>
        <CurrencyLogo hideNetworkLogo currencyInfo={outputCurrencyInfo} size={size} />
      </Box>
      {icon && (
        <Box bottom={-4} position="absolute" right={-4}>
          {icon}
        </Box>
      )}
    </Box>
  )
}
