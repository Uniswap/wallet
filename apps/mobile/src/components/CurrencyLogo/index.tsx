import React from 'react'
import { TokenLogo } from 'src/components/CurrencyLogo/TokenLogo'
import { CurrencyInfo } from 'src/features/dataApi/types'
import { iconSizes } from 'src/styles/sizing'

interface CurrencyLogoProps {
  currencyInfo: NullUndefined<CurrencyInfo>
  size?: number
  hideNetworkLogo?: boolean
}

export const STATUS_RATIO = 2 / 5

export function CurrencyLogo({
  currencyInfo,
  size = iconSizes.icon40,
  hideNetworkLogo,
}: CurrencyLogoProps): JSX.Element | null {
  if (!currencyInfo) return null

  const { currency, logoUrl } = currencyInfo
  const { chainId, symbol } = currency

  return (
    <TokenLogo
      chainId={chainId}
      hideNetworkLogo={hideNetworkLogo}
      size={size}
      symbol={symbol}
      url={logoUrl}
    />
  )
}
