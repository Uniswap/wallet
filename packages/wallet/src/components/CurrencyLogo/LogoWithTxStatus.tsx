import type { IconProps } from '@tamagui/helpers-icon'
import React, { useEffect } from 'react'
import { StyleSheet } from 'react-native'
import { Icons, useTheme } from 'ui/src'
import WalletConnectLogo from 'ui/src/assets/icons/walletconnect.svg'
import MoonpayLogo from 'ui/src/assets/logos/svg/moonpay.svg'
import { Box } from 'ui/src/components/layout'
import { borderRadii } from 'ui/src/theme/borderRadii'
import { CurrencyLogo, STATUS_RATIO } from 'wallet/src/components/CurrencyLogo/CurrencyLogo'
import { TransactionSummaryNetworkLogo } from 'wallet/src/components/CurrencyLogo/NetworkLogo'
import { DappIconPlaceholder } from 'wallet/src/components/WalletConnect/DappIconPlaceholder'
import { ChainId } from 'wallet/src/constants/chains'
import { AssetType } from 'wallet/src/entities/assets'
import { CurrencyInfo } from 'wallet/src/features/dataApi/types'
import { ImageUri } from 'wallet/src/features/images/ImageUri'
import { NFTViewer } from 'wallet/src/features/images/NFTViewer'
import { RemoteImage } from 'wallet/src/features/images/RemoteImage'
import { logger } from 'wallet/src/features/logger/logger'
import {
  NFTTradeType,
  TransactionStatus,
  TransactionType,
} from 'wallet/src/features/transactions/types'
import { WalletConnectEvent } from 'wallet/src/features/walletConnect/types'

interface LogoWithTxStatusProps {
  assetType: AssetType
  txType: TransactionType
  txStatus: TransactionStatus
  size: number
  chainId: ChainId | null
}

interface DappLogoWithTxStatusProps {
  event: WalletConnectEvent
  size: number
  chainId: ChainId | null
  dappImageUrl: string | null
  dappName: string
}

interface CurrencyStatusProps extends LogoWithTxStatusProps {
  assetType: AssetType.Currency
  currencyInfo?: CurrencyInfo | null
}

interface NFTStatusProps extends LogoWithTxStatusProps {
  assetType: AssetType.ERC721 | AssetType.ERC1155
  nftImageUrl?: string
  nftTradeType?: NFTTradeType
}

export function LogoWithTxStatus(props: CurrencyStatusProps | NFTStatusProps): JSX.Element {
  const { assetType, txType, txStatus, size, chainId } = props
  const theme = useTheme()

  const statusSize = size / 2

  const logo =
    txType === TransactionType.FiatPurchase ? (
      <MoonpayLogo height={size} width={size} />
    ) : assetType === AssetType.Currency ? (
      <CurrencyLogo hideNetworkLogo currencyInfo={props.currencyInfo} size={size} />
    ) : (
      <Box
        alignItems="center"
        backgroundColor="$background2"
        borderRadius="$rounded4"
        height={size}
        justifyContent="center"
        overflow="hidden"
        width={size}>
        {props.nftImageUrl && <NFTViewer uri={props.nftImageUrl} />}
      </Box>
    )

  const fill = txStatus === TransactionStatus.Success ? theme.accentSuccess : theme.textSecondary
  const color = theme.background1

  let icon: JSX.Element | undefined
  if (chainId && chainId !== ChainId.Mainnet) {
    icon = <TransactionSummaryNetworkLogo chainId={chainId} size={size * STATUS_RATIO} />
  } else {
    let Icon: React.NamedExoticComponent<IconProps> | undefined
    switch (txType) {
      case TransactionType.Approve:
      case TransactionType.NFTApprove:
        Icon = Icons.Approve
        break
      case TransactionType.Send:
        Icon = Icons.ArrowUpInCircle
        break
      case TransactionType.NFTTrade:
        if (assetType === AssetType.ERC721 || assetType === AssetType.ERC1155) {
          if (props.nftTradeType === NFTTradeType.SELL) {
            Icon = Icons.ArrowUpInCircle
          } else {
            Icon = Icons.ArrowDownInCircle
          }
        }
        break
      // Fiat purchases use the same icon as receive
      case TransactionType.FiatPurchase:
      case TransactionType.Receive:
      case TransactionType.NFTMint:
        Icon = Icons.ArrowDownInCircle
        break
      case TransactionType.Unknown:
        Icon = Icons.QuestionInCircle
        break
    }
    if (Icon) {
      icon = <Icon color={color.get()} fill={fill.get()} height={statusSize} width={statusSize} />
    }
  }

  useEffect(() => {
    if (!icon) {
      logger.warn(
        'statusIcon',
        'GenerateStatusIcon',
        'Could not find icon for transaction type:',
        txType
      )
    }
  }, [icon, txType])

  return (
    <Box alignItems="center" height={size} justifyContent="center" width={size}>
      {logo}
      {icon && (
        <Box bottom={-4} position="absolute" right={-4}>
          {icon}
        </Box>
      )}
    </Box>
  )
}

export function DappLogoWithTxStatus({
  dappName,
  dappImageUrl,
  event,
  size,
  chainId,
}: DappLogoWithTxStatusProps): JSX.Element | null {
  const theme = useTheme()
  const green = theme.accentSuccess.get()
  const yellow = theme.accentWarning.get()
  const fill = theme.background0.get()

  const dappImageSize = size
  const statusSize = dappImageSize * (1 / 2)
  const totalSize = dappImageSize + statusSize * (1 / 4)

  const getStatusIcon = (): JSX.Element | undefined => {
    switch (event) {
      case WalletConnectEvent.NetworkChanged:
        return chainId ? (
          <TransactionSummaryNetworkLogo chainId={chainId} size={statusSize} />
        ) : undefined
      case WalletConnectEvent.TransactionConfirmed:
        return <Icons.Approve color={green} fill={fill} height={statusSize} width={statusSize} />
      case WalletConnectEvent.TransactionFailed:
        return (
          <Icons.AlertTriangle color={yellow} fill={fill} height={statusSize} width={statusSize} />
        )
    }
  }

  const statusIcon = getStatusIcon()

  const fallback = (
    <Box height={dappImageSize}>
      <DappIconPlaceholder iconSize={dappImageSize} name={dappName} />
    </Box>
  )

  const style = StyleSheet.create({
    icon: {
      borderRadius: borderRadii.rounded4,
      height: dappImageSize,
      width: dappImageSize,
    },
    loaderContainer: {
      borderRadius: borderRadii.roundedFull,
      overflow: 'hidden',
    },
  })

  const dappImage = dappImageUrl ? (
    <ImageUri
      fallback={fallback}
      imageStyle={style.icon}
      loadingContainerStyle={{ ...style.icon, ...style.loaderContainer }}
      uri={dappImageUrl}
    />
  ) : (
    fallback
  )

  return statusIcon ? (
    <Box height={totalSize} width={totalSize}>
      <Box left={0} position="absolute" top={0}>
        {dappImage}
      </Box>
      <Box bottom={0} position="absolute" right={0}>
        {statusIcon}
      </Box>
    </Box>
  ) : (
    dappImage
  )
}

/** For displaying Dapp logo with generic WC bade icon */
export function DappLogoWithWCBadge({
  dappImageUrl,
  dappName,
  size,
  chainId,
}: {
  dappImageUrl: string | null
  dappName: string
  size: number
  chainId: ChainId | null
}): JSX.Element {
  const dappImageSize = size
  const statusSize = dappImageSize * STATUS_RATIO
  const totalSize = dappImageSize + statusSize * (1 / 4)
  const dappImage = dappImageUrl ? (
    <RemoteImage
      borderRadius={borderRadii.rounded4}
      height={dappImageSize}
      uri={dappImageUrl}
      width={dappImageSize}
    />
  ) : (
    <DappIconPlaceholder iconSize={dappImageSize} name={dappName} />
  )

  return (
    <Box height={totalSize} width={totalSize}>
      <Box left={2} top={0}>
        {dappImage}
      </Box>
      {chainId && chainId !== ChainId.Mainnet ? (
        <Box bottom={-2} position="absolute" right={-2}>
          <TransactionSummaryNetworkLogo chainId={chainId} size={size * STATUS_RATIO} />
        </Box>
      ) : (
        <Box
          backgroundColor="$background1"
          borderColor="$background0"
          borderRadius="$roundedFull"
          borderWidth={2}
          bottom={-2}
          position="absolute"
          right={-2}>
          <WalletConnectLogo height={statusSize} width={statusSize} />
        </Box>
      )}
    </Box>
  )
}
