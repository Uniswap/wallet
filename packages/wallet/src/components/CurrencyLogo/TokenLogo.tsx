import { Image } from 'react-native'
import { useTheme } from 'tamagui'
import { Box } from 'ui/src/components/layout'
import { Text } from 'ui/src/components/text/Text'
import { iconSizes } from 'ui/src/theme/iconSizes'
import { STATUS_RATIO } from 'wallet/src/components/CurrencyLogo/CurrencyLogo'
import { SHADOW_OFFSET, style, THIN_BORDER } from 'wallet/src/components/CurrencyLogo/styles'
import { ChainId } from 'wallet/src/constants/chains'
import { uriToHttp } from 'wallet/src/utils/uriToHttp'
import { NetworkLogo } from './NetworkLogo'

interface TokenLogoProps {
  url?: string | null
  symbol?: string
  chainId?: ChainId
  size?: number
  hideNetworkLogo?: boolean
}

export function TokenLogo({
  url,
  symbol,
  chainId,
  size = iconSizes.icon40,
  hideNetworkLogo,
}: TokenLogoProps): JSX.Element {
  const theme = useTheme()
  const showNetworkLogo = !hideNetworkLogo && chainId && chainId !== ChainId.Mainnet
  const httpUri = url ? uriToHttp(url)[0] : null

  return (
    <Box alignItems="center" height={size} justifyContent="center" width={size}>
      {httpUri ? (
        <Image
          source={{ uri: httpUri }}
          style={[
            style.image,
            {
              backgroundColor: theme.backgroundOutline.get(),
              width: size,
              height: size,
              borderRadius: size / 2,
              borderColor: theme.backgroundOutline.get(),
              borderWidth: THIN_BORDER,
            },
          ]}
        />
      ) : (
        <Box
          alignItems="center"
          bg="$backgroundOutline"
          borderRadius="$roundedFull"
          height={size}
          justifyContent="center"
          px="$spacing8"
          width={size}>
          <Text
            adjustsFontSizeToFit
            color="$textPrimary"
            flex={0}
            numberOfLines={1}
            textAlign="center">
            {symbol?.slice(0, 3)}
          </Text>
        </Box>
      )}
      {showNetworkLogo && (
        <Box
          bottom={0}
          position="absolute"
          right={0}
          shadowColor="black"
          shadowOffset={SHADOW_OFFSET}
          shadowOpacity={0.1}
          shadowRadius={2}>
          <NetworkLogo chainId={chainId} size={size * STATUS_RATIO} />
        </Box>
      )}
    </Box>
  )
}
