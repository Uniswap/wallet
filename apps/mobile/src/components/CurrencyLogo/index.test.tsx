import React from 'react'
import { arbitrumDaiCurrencyInfo, uniCurrencyInfo } from 'src/test/fixtures'
import { render } from 'src/test/test-utils'
import { CurrencyLogo } from 'wallet/src/components/CurrencyLogo/CurrencyLogo'

jest.mock('ui/src/assets/', () => 'ethereum-logo.png')

it('renders a currency logo without network logo', () => {
  const tree = render(<CurrencyLogo currencyInfo={uniCurrencyInfo} size={20} />)
  expect(tree).toMatchSnapshot()
})

it('renders a currency logo with network logo', () => {
  const tree = render(<CurrencyLogo currencyInfo={arbitrumDaiCurrencyInfo} size={20} />)
  expect(tree).toMatchSnapshot()
})
