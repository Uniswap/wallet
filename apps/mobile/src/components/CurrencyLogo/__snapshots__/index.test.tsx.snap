// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a currency logo with network logo 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "display": "flex",
        "flexBasis": "auto",
        "flexDirection": "column",
        "flexShrink": 0,
        "height": 20,
        "justifyContent": "center",
        "minHeight": 0,
        "minWidth": 0,
        "position": "relative",
        "width": 20,
      },
    ]
  }
>
  <Image
    source={
      {
        "uri": "https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png",
      }
    }
    style={
      [
        {
          "resizeMode": "contain",
        },
        {
          "backgroundColor": "#D2D9EE",
          "borderColor": "#D2D9EE",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "alignItems": "stretch",
          "bottom": 0,
          "display": "flex",
          "flexBasis": "auto",
          "flexDirection": "column",
          "flexShrink": 0,
          "minHeight": 0,
          "minWidth": 0,
          "position": "absolute",
          "right": 0,
          "shadowColor": "rgb(0,0,0)",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 2,
        },
      ]
    }
  />
</View>
`;

exports[`renders a currency logo without network logo 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "display": "flex",
        "flexBasis": "auto",
        "flexDirection": "column",
        "flexShrink": 0,
        "height": 20,
        "justifyContent": "center",
        "minHeight": 0,
        "minWidth": 0,
        "position": "relative",
        "width": 20,
      },
    ]
  }
>
  <Image
    source={
      {
        "uri": "https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png",
      }
    }
    style={
      [
        {
          "resizeMode": "contain",
        },
        {
          "backgroundColor": "#D2D9EE",
          "borderColor": "#D2D9EE",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
</View>
`;
